### PR TITLE
Convert Campaign widgets to entityRef

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Cleans up the code and UI for selecting a campaign. Instead of a select list with a button to load past campaigns, now the entire list of campaigns is searchable and filterable. New campaigns can be created on-the-fly.

Before
----------------------------------------
Ugly message when there are no campaigns.
Button to load past campaigns.
No way to create a new campaign without leaving and reloading the page.

screenshot from 2019-01-22 09-52-53

After
----------------------------------------
New widget can filter out past campaigns and create new ones on-the-fly, so no need for the ugly message or the "show past campaigns" button. The widget also displays the campaign type, status, and dates in the results.

screenshot from 2019-01-21 13-18-52

### CODE REVIEW

**Comment by colemanw** [comment]
@JoeMurray I believe you were involved in creating the "Show past campaigns" button. Can you review this improvement?

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by JoeMurray** [comment]
I'd like to review how this runs in browser, eg the Refine search. From what I can see this is an area where there were opportunities to make a nice usability improvement.

File: templates/CRM/Campaign/Form/addCampaignToComponent.tpl [hunk 1]
```diff
@@ -2,82 +2,20 @@
...
-           'json');
-}
-</script>
-{/literal}
-
-
-{/if}{* add campaign to component if closed.  *}
+  <tr class="{$campaignTrClass}">
+    <td class="label">{$form.campaign_id.label} {help id="id-campaign_id" file="CRM/Campaign/Form/addCampaignToComponent.hlp"}</td>
+    <td class="view-value">{$form.campaign_id.html}</td>
```
Comment (JoeMurray) [inline templates/CRM/Campaign/Form/addCampaignToComponent.tpl]: Just checking: you deliberately are removing |crmAddClass:huge from this html?

Comment (colemanw) [reply templates/CRM/Campaign/Form/addCampaignToComponent.tpl]: Yes. It's not needed for an enityRef.

**Comment by JoeMurray** [comment]
I think commit cbb17d3bbb497b8d07b3e0720c9422e19881715f would be better in its own MR.

**Comment by colemanw** [comment]
> I think commit [cbb17d3](https://github.com/civicrm/civicrm-core/commit/cbb17d3bbb497b8d07b3e0720c9422e19881715f) would be better in its own MR.

Done. #13503

**Comment by JoeMurray** [comment]
I don't know the criteria/process for removing 'AJAX' functions from core. With regular functions in core, there is a fair bit of leeway given the main contract is the API layer. But I imagine developers have been treating anything in an 'AJAX.php' file as fair game for them use, like CRM/Campaign/Page/AJAX.php. If so, then removing functions may require some notice time for deprecation. In particular, I would guess there are extensions out there that allow new entities to be associated with Campaigns, so they might use this function allActiveCampaigns(). Comments @eileenmcnaughton ?

**Review by JoeMurray — APPROVED** [review]

File: CRM/Contact/BAO/Contact.php [hunk 2]
```diff
@@ -3642,4 +3642,16 @@ public static function isFieldHasLocationType($fieldTitle) {
     return FALSE;
   }
 
+  /**
+   * Checks permission to create new contacts from entityRef widget
+   *
+   * Note: other components must return an array of links from this function,
```
Comment (JoeMurray) [inline CRM/Contact/BAO/Contact.php]: 1. Looking in 5.8.0 there were no instantiations of this function. Should this read, '...other components will need to eventually...'? 
2. Is your plan to display create links from Common.js when user has permission to create an entity of that type? 
3. Shouldn't the comment indicate that this is possible for any entity that provides a BAO, whether it is for a component or from an extension?

**Comment by colemanw** [comment]
@JoeMurray I appreciate your attitude of caution about removing the ajax callback, but in this case an api does exist to cover the function - any extension developer wishing to fetch past campaigns can simply use the Campaign api, which has existed for many years. In the (IMO very unlikely) event that someone out there has (a) managed to actually find this obscure function and (b) decided to ignore every piece of advice in our documentation which says avoid internal functions and use the api... I'd have to say they've been warned.

**Comment by JoeMurray** [comment]
> @JoeMurray I appreciate your attitude of caution about removing the ajax callback, but in this case an api does exist to cover the function - any extension developer wishing to fetch past campaigns can simply use the Campaign api, which has existed for many years. In the (IMO very unlikely) event that someone out there has (a) managed to actually find this obscure function and (b) decided to ignore every piece of advice in our documentation which says avoid internal functions and use the api... I'd have to say they've been warned.

I'm betraying my ignorance around what functions in AJAX.php are considered 'callable' by extensions. I'll defer to you.

File: CRM/Contact/BAO/Contact.php (hunk 2 above)
Comment (colemanw) [reply CRM/Contact/BAO/Contact.php]: That function is a new invention in this PR. It's a callback for any BAO to provide links for creating a new item from within the EntityRef widget. Prior to this PR only contacts could do that and the links were hardcoded to be "New Individual" "New Household" and "New Organization".

This is my attempt to add more flexibility and it will probably evolve a little more but I'm mostly happy with it.

File: CRM/Contact/Form/Edit/Address.php [hunk 3]
```diff
@@ -165,11 +165,7 @@ public static function buildQuickForm(&$form, $addressBlockCount = NULL, $sharin
       $form->addElement('checkbox', "address[$blockId][use_shared_address]", NULL, ts('Use another contact\'s address'));
 
       // Override the default profile links to add address form
-      $profileLinks = CRM_Core_BAO_UFGroup::getCreateLinks(array(
-          'new_individual',
-          'new_organization',
-          'new_household',
-        ), 'shared_address');
+      $profileLinks = CRM_Contact_BAO_Contact::entityRefCreateLinks() ? CRM_Core_BAO_UFGroup::getCreateLinks('', 'shared_address') : FALSE;
```
Comment (JoeMurray) [inline CRM/Contact/Form/Edit/Address.php]: This cleanup seems to make use of new function but not be related to Campaigns. Should it be in a different MR to make things easier to review and approve?

**Review by JoeMurray — APPROVED** [review]

**Comment by JoeMurray** [comment]
This MR improves the UX. It replaces many lines of code with a small number of more elegant ones that provide a path for similar cleanup elsewhere. 

I tested the campaign widget on Add Activity and Edit Activity and Add Contribution and all worked properly. I tested the filtering options and all worked nicely.

The create links for adding new contacts on the Address sharing widget were not properly filtered. As mentioned above I'd prefer to see that moved to a different PR.

If that is fixed or moved out of this MR then I'd say this is ready to merge.

**Comment by colemanw** [comment]
Thanks Joe. What problem did you run into with the shared address widget?

**Comment by JoeMurray** [comment]
My comment above is still pending, I think due to them being on a line removed by a force push from Coleman. Here they are repeated.

Removing CiviCRM: add contacts perm from all role, logging in as demo user (as opposed to admin user 1 which has all permissions), then editing a contact, I see


However, as expected, clicking on the New Individual, New Organization, and New Household links, an error is displayed:


Could you fix so that the create links are not displayed for users without the correct permissions? It might be appropriate to move this to a different MR.

**Comment by colemanw** [comment]
Thanks @JoeMurray - I just tested and found the problem is preexisting and not related to this PR. So I'll open up a new PR to solve that issue and merge this based on your review. Thanks!

**Comment by eileenmcnaughton** [comment]
@colemanw I've deployed this & I'm not loving it - I think it was totally broken when I first used it, but then I realised I could find my campaign

I think the issue is the fact there is no pre-population of a list so it seems like none exist - I'm pretty sure there is a param for that that you turn on & off in other entity references to make it feel a bit more like a select when you load it?

**Comment by colemanw** [comment]
@eileenmcnaughton yes we could enable that mode. However it would make the filters harder to find (they'd only appear if you get no search result).

**Comment by eileenmcnaughton** [comment]
@colemanw hmm - I guess what people will find more convenient will depend on how many campaigns they have. However, it would be tough to use for our org by anyone other than me - ie. I know what campaign I'm gonna search for but I think others wouldn't have a clue until they see a list.
